### ISSUE TITLE
Ordering of Include-Resource appears to be dependent on file system, affecting reproducibility

### ISSUE THREAD

**Issue by diabonas** [bndtools/bnd#4221]
I am investigating a reproducibility issue in Apache Commons Lang, see https://github.com/apache/commons-lang/pull/578#issuecomment-656022221: even with `-noextraheaders: true` set as per the [Bnd README](https://github.com/bndtools/bnd/blob/cc61fa32da6b779a9607d02b78c2052015493eb6/maven/bnd-maven-plugin/README.md#reproducible-builds), different builds still generate different JAR files when built on different file systems. When building on a Btrfs file system, the `Include-Resource` header for this project is
```
Include-Resource: META-INF/LICENSE.txt=LICENSE.txt,META-INF/NOTICE.txt=NOTICE.txt
```
while building on a tmpfs yields
```
Include-Resource: META-INF/NOTICE.txt=NOTICE.txt,META-INF/LICENSE.txt=LICENSE.txt
```
For reproducible builds it would be helpful to sort the entries in the header in a filesystem- and locale-independent way, e.g. according to the C locale, or really anything guaranteeing a consistent sorting.

**Comment by rotty3000** [comment]
You should replace the usage of `Include-Resource` (which is a header and will persist in the manifest) with `-includeresource` (which is an instruction rather, and will not persist in the manifest but evokes the same functionality.)

**Comment by bjhargrave** [comment]
> You should replace the usage of `Include-Resource` (which is a header and will persist in the manifest) with `-includeresource`

Yes.

`Include-Resource` was the original way, now superseded by `-includeresource`.

There is no real reason that you should have `Include-Resource` in your generated manifest.

**Comment by bjhargrave** [comment]
In looking at the pom in the commons-lang project, you appear to be using `maven-bundle-plugin` (which is not from this Bnd project) and also do not appear to be setting `Include-Resource` yourself. Perhaps it is `maven-bundle-plugin` which is setting `Include-Resource` as part of its default behavior (to always include LICENSE and NOTICE files)?  If that is the case, then you may need to get them to fix to either (1) switch to use `-includeresource` or (2) consistently sort the value of the `Include-Resource` header. I can find no code in Bnd which would not preserve the order of the specified `Include-Resource` header value.

So, in summary, I do not believe there is any bug in Bnd. Any bug here would exist in `maven-bundle-plugin`.

Note, you could switch to using `bnd-maven-plugin` (from this Bnd project), but you will need to configure things differently than `maven-bundle-plugin` since `bnd-maven-plugin` does not have all the default "magical" configuration of `maven-bundle-plugin`.

**Comment by bjhargrave** [comment]
As a final note, you could specify `-removeheaders: Include-Resource` to tell Bnd to elide the `Include-Resource` header from the generated manifest to sidestep the issue with `maven-bundle-plugin`.

**Comment by diabonas** [comment]
Thank you very much for the information, I will try to get `Include-Resource` replaced by `-includeresource` in `maven-bundle-plugin` then!

If `Include-Resource` is generally not a useful thing to have in the manifest, would it make sense to remove it from Bnd entirely, perhaps conditionally if `-noextraheaders: true` is set? I imagine there are a lot of legacy projects, not only `maven-bundle-plugin`, which still use it and are therefore potentially unreproducible?

**Comment by bjhargrave** [comment]
`-removeheaders: Include-Resource` is the way to remove it from the manifest. In fact, this is in the default `maven-bundle-plugin` configuration:

https://github.com/apache/felix-dev/blob/1a573e93f797b5a281e72c079a1b519c509c3f18/tools/maven-bundle-plugin/src/main/java/org/apache/felix/bundleplugin/BundlePlugin.java#L1807-L1808

**Comment by rotty3000** [comment]
I commented on the suggested commons-lang fix that the syntax used is incorrect https://github.com/apache/commons-lang/pull/578/files#r452371269

**Comment by garydgregory** [comment]
> `-removeheaders: Include-Resource` is the way to remove it from the manifest. In fact, this is in the default `maven-bundle-plugin` configuration:
> 
> https://github.com/apache/felix-dev/blob/1a573e93f797b5a281e72c079a1b519c509c3f18/tools/maven-bundle-plugin/src/main/java/org/apache/felix/bundleplugin/BundlePlugin.java#L1807-L1808

The better solution would be to offer a PR to the `maven-bundle-plugin` to sort `Include-Resource` entries at the point where these are gathered. I'm not quite sure where in the plugin that happens though.

**Comment by rotty3000** [comment]
I am suspicious that when you set this:
```
 <_removeheaders>Bnd-LastModified</_removeheaders>...
```
you inadvertently overloaded the default value that @bjhargrave pointed out in the plugin.

**Comment by rotty3000** [comment]
Instead of setting `<_removeheaders>Bnd-LastModified</_removeheaders>...` try
`<_noextraheaders>true</_noextraheaders>`

**Comment by diabonas** [comment]
> I am suspicious that when you set this:
> 
> ```
>  <_removeheaders>Bnd-LastModified</_removeheaders>...
> ```
> 
> you inadvertently overloaded the default value that @bjhargrave pointed out in the plugin.

The `Include-Resource` header is definitely present even without the `<_removeheaders>` patch, see the first diff in https://github.com/apache/commons-lang/pull/578#issuecomment-656022221. The only thing I could find that might be causing this is [this entry in the commons-parent `pom.xml`](https://github.com/apache/commons-parent/blob/b39108aba84b93b5b57b8bdc6c934fbadd98521f/pom.xml#L902), could that be an issue?

**Comment by rotty3000** [comment]
> The only thing I could find that might be causing this is [this entry in the commons-parent `pom.xml`](https://github.com/apache/commons-parent/blob/b39108aba84b93b5b57b8bdc6c934fbadd98521f/pom.xml#L902), could that be an issue?

Definitely!

**Comment by rotty3000** [comment]
With that, if you still _really_ must remove even more headers (be careful mind you) you _can_ set additional remove headers without breaking either maven-bundle-plugin's defaults or the parent pom's by using _merged properties_ syntax.

`<_removeheaders.commons-lang>Foo, Bar-Baz</_removeheaders.commons-lang>`

**Comment by diabonas** [comment]
Thanks again! Just to be clear, so the correct solution would be to use something like https://github.com/diabonas/commons-parent/commit/8795f4457c3c87a1df09d83d9b2d1462272dac9f (use `_noextraheaders` and replace `_removeheaders` by `_removeheaders.commons-parent` to avoid overwriting the default list of headers to be removed)?

**Comment by bjhargrave** [comment]
Using `-removeheaders.commons-lang` as an instruction supplements the `-removeheaders` instruction in the parent POM. So it is a good choice (unless you need to override the `-removeheaders` instruction in the parent POM.)

**Comment by diabonas** [comment]
Cheers, I am going to close this issue then since the solution will be to fix either `commons-parent` or `maven-bundle-plugin`, no action needed regarding Bnd.

**Comment by bjhargrave** [comment]
> When building on a Btrfs file system, the `Include-Resource` header for this project is
> 
> ```
> Include-Resource: META-INF/LICENSE.txt=LICENSE.txt,META-INF/NOTICE.txt=NOTICE.txt
> ```
> 
> while building on a tmpfs yields
> 
> ```
> Include-Resource: META-INF/NOTICE.txt=NOTICE.txt,META-INF/LICENSE.txt=LICENSE.txt
> ```
> 
> For reproducible builds it would be helpful to sort the entries in the header in a filesystem- and locale-independent way, e.g. according to the C locale, or really anything guaranteeing a consistent sorting.

This ordering issue may be actually caused by Maven itself. 

https://github.com/apache/commons-parent/blob/b39108aba84b93b5b57b8bdc6c934fbadd98521f/pom.xml#L423-L433

adds these resources to the Maven Project. `maven-bundle-plugin` just processes the project resources as supplied by Maven

https://github.com/apache/felix-dev/blob/016acaeb667bbba96127d3e371ecf3712aa839bb/tools/maven-bundle-plugin/src/main/java/org/apache/felix/bundleplugin/BundlePlugin.java#L2032

So if Maven returns the resources in different orders for different file systems, neither `maven-bundle-plugin` nor Bnd at at fault (although, as the generated of the resource instruction, `maven-bundle-plugin` could sort the elements of the value). So the simplest and quickest solution is to elide the `Include-Resource` header from the manifest.

**Comment by diabonas** [comment]
> This ordering issue may be actually caused by Maven itself.

Yikes, I wonder whether this might cause reproducibility issues in other places as well... Thanks for taking the time to dig deeper!

> So the simplest and quickest solution is to elide the `Include-Resource` header from the manifest.

Given that this would be the `maven-bundle-plugin` default anyway that is unluckily overwritten by `commons-parent`, I think fixing the latter so that `Include-Resource` is removed once again is the most pragmatic thing to do here.
